Source/SwiftLintFramework/Rules/Lint/InvalidSwiftLintCommandRule.swift

```diff
@@ -0,0 +1,126 @@
+import Foundation
+import SwiftSyntax
+
+struct InvalidSwiftLintCommandRule: SwiftSyntaxRule, ConfigurationProviderRule {
+    var configuration = SeverityConfiguration(.warning)
+
+    init() {}
+
+    static let description = RuleDescription(
+        identifier: "invalid_swiftlint_command",
+        name: "Invalid SwiftLint Command",
+        description: "swiftlint command does not have a valid action or modifier",
+        kind: .lint,
+        nonTriggeringExamples: [
+            Example("// swiftlint:disable unused_import"),
+            Example("// swiftlint:enable unused_import"),
+            Example("// swiftlint:disable:next unused_import"),
+            Example("// swiftlint:disable:previous unused_import"),
+            Example("// swiftlint:disable:this unused_import")
+        ],
+        triggeringExamples: [
+            Example("// ↓swiftlint:"),
+            Example("// ↓swiftlint: "),
+            Example("// ↓swiftlint::"),
+            Example("// ↓swiftlint:: "),
+            Example("// ↓swiftlint:disable"),
+            Example("// ↓swiftlint:dissable unused_import"),
+            Example("// ↓swiftlint:enaaaable unused_import"),
+            Example("// ↓swiftlint:disable:nxt unused_import"),
+            Example("// ↓swiftlint:enable:prevus unused_import"),
+            Example("// ↓swiftlint:enable:ths unused_import"),
+            Example("// ↓swiftlint:enable"),
+            Example("// ↓swiftlint:enable:"),
+            Example("// ↓swiftlint:enable: "),
+            Example("// ↓swiftlint:disable: unused_import")
+        ]
+    )
+
+    func makeVisitor(file: SwiftLintFile) -> ViolationsSyntaxVisitor {
+        Visitor(viewMode: .sourceAccurate)
```

Comment:
Can we just use `file.commands` instead of a custom visitor?

Reply:
I don't that will work, as we're looking for invalid commands, that would have failed parsing, so they won't end up in `file.commands`

Reply:
See this small example test case:
```swift
class CommandVisitorTest: XCTestCase {
    func test() {
        let file = SwiftLintFile(contents: """
            // swiftlint:disable:next foo bar
        """)

        XCTAssertEqual(file.commands.map(\.ruleIdentifiers), [[.single(identifier: "foo"), .single(identifier: "bar")]])
    }
}
```
It shows that even unknown identifiers are parsed and collected.

Reply:
So the `invalid_swiftlint_command` will not actually pick up invalid rule identifiers, but `superfluous_disable` already reports those:

`Superfluous Disable Command Violation: 'x' is not a valid SwiftLint rule. Please remove it from the disable command. (superfluous_disable_command)`

`invalid_swiftlint_command` picks up some cases that will fail parsing in the CommandVisitor, and never end up in `file.commands`

```
class CommandVisitorTest: XCTestCase {
    func testInvalidActions() {
        let file = SwiftLintFile(contents: """
            // swiftlint:dissable:next foo bar
            // swiftlint:
            // swiftlint:
            // swiftlint::
            // swiftlint:disable
            // swiftlint:dissable unused_import
            // swiftlint:enaaaable unused_import
            // swiftlint:enable
            """)
        XCTAssertTrue(file.commands.isEmpty)
    }
}
```

In fact, it seems really dodgy that something like

`// swiftlint:disable:nxt function_parameter_count`

end up in `file.commands`, as the invalid modifier will be ignored there, and then `function_parameter_count` will be disabled for the rest of the file, which is surely not what the user intended.

This rule will at least help flag those cases up to the user.

Reply:
I've actually updated Command now, so that it's initialiser will return nil if the modifier is invalid, so 

`// swiftlint:disable:nxt unused_import` 

will have no effect, instead of disabling `unused_import` for the rest of the file. This seems more reasonable than the current behaviour, and makes the rule code a lot simpler, because previously I had a lot of code just for detecting and parsing the modifier to see if it was bad.

Reply:
```

class CommandVisitorTest: XCTestCase {
    func testInvalidActions() {
        let file = SwiftLintFile(contents: """
            // swiftlint:dissable:next foo bar
            // swiftlint:
            // swiftlint:
            // swiftlint::
            // swiftlint::
            // swiftlint:disable
            // swiftlint:dissable unused_import
            // swiftlint:enaaaable unused_import
            // swiftlint:enable
            // swiftlint:enable:
            // swiftlint:enable:
            // swiftlint:disable: unused_import
            // swiftlint:disable:nxt unused_import
            // swiftlint:enable:prevus unused_import
            // swiftlint:enable:ths unused_import
            """)
        XCTAssertTrue(file.commands.isEmpty)
    }
}
```

Reply:
Okay, I got that. Thank you for the clarification.

I still think `CommandVisitor` should parse these commands once and for all and the newly added rule just evaluates all commands parsed from a file (`file.commands`). That can be achieved by extending `Command.Action` and `Command.Modifier` with `.invalid(String)` enum cases. We save another parsing pass and partial duplication of code in that way. It boils down to validating the available commands.

Does that make sense or am I missing something again?

Reply:
Yep, I totally agree one pass would be better and even less code.

It does seem a bit mean to leave invalid commands in `file.commands` though, and make callers check to see if they are valid or not.

How about something like this: 

Parse all the commands in CommandVisitor, and valid or not (including those cases which would fail parsing today), store the results in the `commandsCache`, and give `Command` an `isValid` property (which can look at the modifier and action to see if they're valid). There's also a case where no rules are supplied, but I guess I can always check the number of ruleIdentifiers for that.

Define `commands` as follows:

`internal var commands: [Command] { commandsCache.get(self).filter { $0.isValid } }`

and add `invalidCommands` defined as

`internal var invalidCommands: [Command] { commandsCache.get(self).filter { !$0.isValid } }`

That way everything is parsed just once, but callers of `command` can reasonably expect that those commands are valid.

Reply:
Something like this:

https://github.com/realm/SwiftLint/compare/main...mildm8nnered:SwiftLint:mildm8nnered-invalid-swiftlint-directives-rule-one-pass?expand=1

Reply:
Yes, I like it more! Would you like to apply it to this PR for review?

Reply:
Sure - just merged those changes into this branch!

Reply:
Any thoughts on this @SimplyDanny ?